Improved Game State Detection and Support for Logging Abilities

PoizenJam opened the pull request:
Updated RockSniffer Lib to support better logging and game-state detection:

1. Added GAME_STATE detection to SnifferState track whether song is (or has been) paused (separate from game stage!)
2. Added songInfo triggers to Song Start and End for the purposes of tracking playthrough history in logs
3. Improved tuning dictionary, including fallback logic to raw string tuning if no valid definition found
4. Improved handling of tuning display for bass (i.e., ignore 'B and e' string)
5. Improved VocalDetails Handling

PoizenJam committed:
-Changed timestamps in Logger.cs to 24-hour format with milliseconds; preferred format when used for video editing
-Enhanced the playback tracking functionality in Sniffer.CS, allowing Rocksniffer to track pauses and restarts
-Added new logger outputs on Song Start and Song End, which output crucial information about the song and the user's performance to the log. This can be used to extract a record of performance / use to extract clips from a larger recording.
-Added a SONG_PAUSED state to SnifferState

PoizenJam committed:
Rocksniffer_PJ 0.4.2 is a small feature update:
-If tuning tuning match in dictionary is not found, it will fall back on raw string tuning; i.e., if 'E Standard' was not found, report 'EADGBE' instead of 'Custom Tuning'
-If using bass, tuning name will be trimmed to four characters

PoizenJam committed:
Minor Update:
-Invert tuning dictionary and
-add string trim to 'author' log line to remove leading spaces

PoizenJam committed:
-Set to ignore A220 offsets and normalize to A440 for any weird A220 offsets

PoizenJam committed:
Major update:
-Migrated playthrough logginsd functions from RockSnifferLib to RockSniffer_PJ
-Added SQL and CSV playthrough tracking abilit and supporting functions
-Brought version in line with updated RocksnifferLib features, including support for new editions of Rocksmith

PoizenJam committed:
Minor update:
-Fix SONG_ENDING trigger

PoizenJam committed:
Minor update to Sniffer.cs: improve song ending detection logic to avoid false positives

PoizenJam committed:
Added ArrangementPath and ArrangementTuning information to tracked information for playthrough logging.

PoizenJam committed:
-Improve game_state detection logic to fix floating point handling errors that sometimes occur with different .net versions

PoizenJam committed:
-UPDATE: Improve algorithm for detecting song pauses and restarts; should reduce false negatives and false positives aline.

PoizenJam committed:
Pause detection: bump stall threshold to 5, add end-of-song guard

PoizenJam committed:
Custom tuning fallback: show raw string notes, truncate to 4 for bass

PoizenJam committed:
-Merge with upstream commits to main fork

PoizenJam committed:
-Remove duplicate variable declaration

PoizenJam committed:
v0.6.5: validate arrangementID memory reads (format + cross-reference)

PoizenJam committed:
v0.6.5: validate arrangementID memory reads (format + cross-reference)

PoizenJam committed:
Merge commit '76aa083'

PoizenJam committed:
v0.6.5-hotfix: nonstopplay arrangement context + browse-spam fixes

- Pass arrangement context through OnActualSongEnd args
- Silent cross-reference clear (no warning spam)
- gameStage detection for Nonstop nonstopplaygame<->nonstopplayhub
- Deferred LogSongStart waits ~5s for arrangementID memory to populate
- Force-end old song on songID change before currentCDLCDetails update
- Fire-once guards prevent double start/end firing
- RSMemoryReader: reset arrangementID on songID change, always propagate to prevReadout

PoizenJam committed:
v0.6.5 hotfix2: replay/restart guard reset + prev-path heuristic

PoizenJam committed:
v0.6.5 hotfix3: initTime guard + persistent fallback chain

PoizenJam committed:
v0.6.5 hotfix4: plumb wasNonstopMode through events

PoizenJam committed:
v0.6.5 hotfix5: Path memory pointer integration; remove deferral

PoizenJam committed:
v0.6.5 hotfix5.1: first-match-wins in Path resolution

PoizenJam committed:
v0.6.5 hotfix5.2: fix Path pointer Remastered base address

PoizenJam committed:
v0.6.5: remove vestigial fallback chain (SnifferRuntimeState, defaultArrangementType, lastResolvedPath)

PoizenJam committed:
gitattributes: pin LF on checkout (no behavior change)

PoizenJam committed:
v0.6.6: gameStage migrated to static address (Rocksmith2014.exe+0xF5F7C9)

Replaces the pointer chain (entry 0x00F5C5AC + edition shift, offsets
[0x18, 0x18, 0xC, 0x14]) with a direct read at the canonical .data-section
gameStage buffer. The chain returned garbage in SA song-select / song-options
/ tuner and silently dropped las_pause / nsp_pause / nsp_tuner stages
entirely. The static address tracks all observed states correctly.

Known engine behavior (intentionally not corrected at the reader layer):
gameStage does NOT update on pause-resume or pause-restart for any mode.
Consumers needing actual play/pause state should use game_state, which
derives play/pause via timer-stall detection in Sniffer.UpdateState().

Discovery via Cheat Engine string-scan for 'gcpre'; canonical writer
narrowed from 13 hits by observing round-trip behavior across mode/menu
transitions. Beta and Learn_And_Play offsets back-derived using the
+0x3080 / +0x4080 convention used throughout MemoryOffsets.cs (untested
on those editions; consistent with the v0.6.5 hotfix5 currentPath caveat).

PoizenJam committed:
renormalize line endings to LF (no behavior change)

PoizenJam committed:
v0.6.7: flag-driven pause detection (Rocksmith2014.exe+0xF5F5FC)

Adds direct read of Rocksmith's pause-menu mode byte at module+0xF5F5FC
(Remastered), exposed as RSMemoryReadout.pauseMenuMode (raw) and
isPaused (derived bool). Cross-mode validated for SA, LaS, NSP, and
Guitarcade minigames; verified to survive game relaunch as a
.data-section static.

Replaces the prior timer-stall heuristic in SnifferState.UpdateState
with first-poll instant pause entry AND pause exit. The tuner-from-
pause edge case is handled cleanly without special-case logic:
the engine transitions pauseMenuMode between 2 (pause menu) and
1 (tuner sub-overlay) without dropping to 0, so isPaused stays true
throughout and SONG_PAUSED is preserved correctly.

Also adds StringEnumConverter to RSMemoryReadout.mode for readable
JSON serialization (mode now appears as 'SCOREATTACK' instead of 2).

Dead code removed: STALL_THRESHOLD, STALL_EPSILON,
END_OF_SONG_PAUSE_GUARD constants and stallCount field, plus all
six stallCount reset lines across the file. lastObservedTimer
retained for diagnostic logging only.

Discovery: cross-referencing Koko's named candidates at +F5F545
(MustBlockInputsDueToPauseMenu) and +F5F5DB (EnablePauseMenu) -
both dead in current Remastered build - led to inspection of
nearby memory and discovery of the working byte at +F5F5FC.

Beta and Learn_And_Play offsets back-derived using the standard
+0x3080 / +0x4080 convention; not independently verified.

PoizenJam committed:
v0.6.7: state-machine guards on IN_MENUS and SONG_STARTING transitions

Adds gameStage-based guards to prevent the state machine from getting
stuck in SONG_STARTING in three scenarios that previously misbehaved:

  - RockSniffer attaches to an already-running Rocksmith at main menu
  - Rocksmith restarts while RockSniffer is running
  - User presses Esc on a song's loading screen before timer advances

Two new private static helpers added:
  IsPotentiallyPlayingGameStage(string) - allowlist gating IN_MENUS exit
  IsDefinitelyMenuGameStage(string)     - menu/tuner/songreview detector
                                          for SONG_STARTING escape hatch

Both deliberately exclude *_pause gameStages, which are sticky after
first pause during any song and would otherwise produce false signals.

IN_MENUS -> SONG_SELECTED now requires gameStage to be one of:
  las_game, sa_game, nonstopplaygame, las_pause, sa_pause, nsp_pause

SONG_STARTING now has a new escape branch firing when gameStage
transitions to mainmenu, gcpre, learnasong, scoreattack, nonstopplay,
nsp_main, las_songs, las_options, las_tuner, nsp_tuner, las_songreview,
sa_songreview, panel_bib, shop, or gcade.

Unknown gameStages default to NOT triggering transitions in either
direction (conservative — new menu stages we encounter later can be
added to IsDefinitelyMenuGameStage as observed).

No changes outside Sniffer.cs. Program.cs version constant unchanged
(still 0.6.7) since v0.6.7 has not been published publicly yet and
this is a fold-in to the existing in-flight release.

PoizenJam committed:
v0.6.7: disambiguate pause-resume vs pause-exit via songTimer snapshot

When the user paused a song and then hit Exit from the pause menu,
the state machine briefly flickered through SONG_PLAYING before
landing on IN_MENUS. Root cause: pauseMenuMode (and therefore
isPaused) flips to 0 the moment Exit is pressed, while songTimer
stays frozen at the pause-point until the engine completes the
transition to menus. The v0.6.7 flag-driven resume condition
(!isPaused && songTimer > initTime) was satisfied during the
1-2 polls between Exit-pressed and songTimer-dropping-to-0.

Fix: capture songTimer at SONG_PAUSED entry as pauseTimerSnapshot,
and add a third clause to the resume condition requiring
songTimer != pauseTimerSnapshot. Real resume → Rocksmith re-advances
the timer (possibly with a small rewind for catch-up) → snapshot
mismatch → resume fires. Exit → songTimer stays frozen at the
pause-point → snapshot matches → resume condition stays false →
the existing songTimer <= initTime branch catches the IN_MENUS
transition cleanly when the timer eventually drops to 0.

pauseTimerSnapshot is reset at all six song-state-reset sites
alongside lastObservedTimer (matching the existing reset pattern)
plus on successful resume.

No other changes. Tuner-from-pause behavior preserved
(pauseMenuMode 2→1 transition keeps isPaused true; resume condition
isn't evaluated during tuner sub-flow). All other v0.6.7 work
unchanged.

PoizenJam committed:
v0.6.8: PLAY_arrID chain for Nonstop arrangement-ID resolution

PoizenJam committed:
v0.6.8: PLAY_arrID chain for Nonstop arrangement-ID resolution

PoizenJam committed:
v0.6.8: gameStage-derived mode classification

- RSMode enum expanded (NONSTOPPLAY, GUITARCADE, SESSION, LESSONS, MENU)
- DeriveModeFromGameStage classifier with exact + prefix matching
- Bare 'tuner' gameStage persists prior mode
- Legacy note-data-driven mode writes removed

PoizenJam committed:
v0.6.8: always-propagate readout.mode to prevReadout

DoReadout returns prevReadout, not readout. The mode derivation writes
to readout.mode, but prevReadout.mode is only refreshed in-song via the
songTimer-gated CopyTo. In menus, prevReadout.mode retained the last
in-song value (e.g. SCOREATTACK after a Score Attack session, visible
even on the main menu). Add mode to the always-propagate block alongside
gameStage / currentPath / pauseMenuMode, since v0.6.8 mode is meaningful
in every gameStage.

PoizenJam committed:
v0.6.8 audit pass: MP gate + propagation + stale-comment cleanup

- OnActualSongStartArgs/OnActualSongEndArgs: add wasMultiplayerMode field
- Sniffer.cs: capture currentSongRunWasMultiplayerMode at LogSongStart
  from currentMemoryReadout.mode == RSMode.MULTIPLAYER; propagate through
  both event invocations; reset alongside wasNonstopMode
- Sniffer.cs: update stale 'known unfixable' Nonstop arrangement-ID
  comment to reflect v0.6.8 PLAY_arrID unlock

PoizenJam committed:
v0.6.9: completion-flag helper for force-end paths

Replaces maxTime-near-songLength heuristic with three deterministic
signals: pause-driven exit (currentState==SONG_PAUSED), no-input boot
(maxTime<0.1s), and SA 3-strike fail (FailedPhrases>=3). SA check is
first so last-phrase fails still resolve to completed=false.

Fixes two prior false-negative sources: natural completion that missed
the SONG_ENDING transition due to poll-cadence race, and NSP transitions
where maxTime ended fractionally below songLength-0.5s.

PoizenJam committed:
v0.6.9: deterministic completion-flag + pauseMenuMode transition gate

Sniffer.cs: two related changes for v0.6.9.

1. DetermineCompletedForForceEnd() helper using three deterministic
   signals to decide the completed flag at SONG_PLAYING->timer=0,
   songID-change force-end, and gameStage force-end:
     - currentState == SONG_PAUSED (case 1: pause-driven exit/restart/skip)
     - maxTime < 0.1s (case 2: no-input boot at song start)
     - SA FailedPhrases >= 3 (case 3: 3-strike fail, checked FIRST so a
       last-phrase fail still resolves to false)
   Replaces the pre-v0.6.9 maxTime>=songLength-0.5f heuristic at all
   three force-end sites. Fixes two prior false-negative sources:
   natural completion that missed the SONG_ENDING transition due to
   poll-cadence race, and NSP transitions where maxTime ended
   fractionally below the tolerance.

2. previousPauseMenuMode field gating pause-entry on the 0->non-zero
   TRANSITION rather than raw current state. Fixes false
   'Song Paused!'/'Song Resumed!' pair immediately after restart-from-
   pause: Rocksmith can briefly keep pauseMenuMode reading the previous
   run's value for several polls into the new song before clearing it.
   Tuner-from-pause (2->1) unaffected since we're already in SONG_PAUSED.

paused field semantics unchanged (stays sticky across pause->resume).

PoizenJam committed:
v0.6.9: deterministic completion-flag for force-end paths

Adds DetermineCompletedForForceEnd() helper using three deterministic
signals to decide the completed flag at SONG_PLAYING->timer=0, songID-
change force-end, and gameStage force-end:

  1. currentState == SONG_PAUSED (case 1: pause-driven exit/restart/skip,
     detected because UpdateState() hasn't run yet when force-end fires).
  2. maxTime < 0.1s (case 2: no-input boot at song start, before the
     timer can advance past the initialization flash threshold).
  3. SA FailedPhrases >= 3 (case 3: 3-strike fail, checked FIRST so a
     last-phrase fail still resolves to false even when maxTime is
     near songLength).

Replaces the pre-v0.6.9 maxTime>=songLength-0.5f heuristic at all three
force-end sites. Fixes two prior false-negative sources: natural
completion that missed the SONG_ENDING transition due to poll-cadence
race, and NSP transitions where the previous song's maxTime ended
fractionally below the tolerance.

paused field semantics unchanged (stays sticky across pause->resume,
for SQL/CSV/event-log columns).

kokolihapihvi commented on Events/OnActualSongEndArgs.cs:
Why do we need "Actual" events instead of expanding the existing ones?
Since you're only adding fields, it should be backward compatible.

PoizenJam replied:
You're right that as start/end signals the two pairs are redundant. I'm on board with converging instead of shipping a parallel pair.

One issue before I do it- The end side is clean, but old OnSongStarted still does one thing the _Actual_ events don't: it fires at song-select, before gameplay vegins, and stamps the "metadata loaded" timestamp my history uses for the selection-time column, which is separate from the gameplay start and end timestamps.

So how about this: drop OnSongEnded, make OnActualSongStart/OnActualSongEnd the new OnSongStart / OnSongEnded, and keep the early event but rename it to what it is: a metadata-loaded or song-selected signal. That kills the redundancy without losing the selection timestamp. Sound good?

kokolihapihvi commented on Events/OnActualSongEndArgs.cs:
```diff
@@ -0,0 +1,36 @@
+        // Captured at song START in Sniffer.cs and preserved through end (via the run-context
+        // fields). Pre-v0.6.8 gated playthrough writes; v0.6.8 lifted that gate once
+        // PLAY_arrID resolved Nonstop arrangement-ID reliably. Field preserved for
+        // any downstream consumer that wants the contextual flag.
```
We don't need version history in the comments.

PoizenJam replied:
Cleaned up in https://github.com/PoizenJam/RockSnifferLib/commit/b338cfd5c24082d137dfc3b509c4ffb89418b2ed

kokolihapihvi commented on Events/OnActualSongStartArgs.cs:
Same here, why not extend the existing events instead of adding new "Actual" events?

PoizenJam replied:
See previous comment, copied for reference:

You're right that as start/end signals the two pairs are redundant. I'm on board with converging instead of shipping a parallel pair.

One issue before I do it- The end side is clean, but old OnSongStarted still does one thing the _Actual_ events don't: it fires at song-select, before gameplay vegins, and stamps the "metadata loaded" timestamp my history uses for the selection-time column, which is separate from the gameplay start and end timestamps.

So how about this: drop OnSongEnded, make OnActualSongStart/OnActualSongEnd the new OnSongStart / OnSongEnded, and keep the early event but rename it to what it is: a metadata-loaded or song-selected signal. That kills the redundancy without losing the selection timestamp. Sound good?

kokolihapihvi commented on Events/OnActualSongStartArgs.cs:
```diff
@@ -0,0 +1,29 @@
+        // start. Pre-v0.6.8 used to gate playthrough tracking — that gate was lifted
+        // in v0.6.8 once PLAY_arrID made arrangement resolution reliable in Nonstop.
+        // The field is preserved for any downstream consumer that wants the
+        // contextual flag.
```
Don't need version history in the comments, the first line is probably enough to describe this field and its purpose.

PoizenJam replied:
Cleaned up in https://github.com/PoizenJam/RockSnifferLib/commit/b338cfd5c24082d137dfc3b509c4ffb89418b2ed

kokolihapihvi commented on Logging/Logger.cs:
```diff
@@ -81,11 +86,34 @@ public static void LogException(Exception e)
+            string message = string.Format(pattern, p);
 
-            Console.WriteLine(pattern, p);
+            // Fire events for EVENT=START and EVENT=END with timestamp and message
+            // These fire BEFORE any output, ensuring timestamp accuracy
+            if (message.StartsWith("EVENT=START"))
+            {
+                OnEventStartLogged?.Invoke(null, new EventLoggedArgs { 
+                    Timestamp = timestamp, 
+                    Message = message 
+                });
+                // Suppress output - EventLogger will handle it based on config
+                return;
+            }
+            else if (message.StartsWith("EVENT=END"))
+            {
+                OnEventEndLogged?.Invoke(null, new EventLoggedArgs { 
+                    Timestamp = timestamp, 
+                    Message = message 
+                });
+                // Suppress output - EventLogger will handle it based on config
+                return;
+            }
```
This seems pretty hacky, should probably come up with a better way to log events, and write them  separately to a 'events.log' file or similar, instead of mixing in with the rest of the log.

Friedfisch replied:
The Log function for sure is the wrong place to fire events. Log might listen to events. Otherwise this is for sure some unwanted side effect.

PoizenJam replied:
This is explicitly vestigal; a bodged solution from 2024, before I got the csv / dB functionality working. I simply dumped the playthrough history to sniffer.log as semicolon delimited keys and parsed it to CSV using a separate python script.

All to say, yes, it ought to be deprecated for the main fork. I'll put it on the list.

PoizenJam replied:
Mostly addressed- just pending additional discussion about potential events.log reformulation

kokolihapihvi commented on RSHelpers/MemoryOffsets.cs:
```diff
@@ -79,18 +79,65 @@ public static (int entryAddress, int[] offsets) GetArrangementHashPointer(RSEdit
+    /// module+0xF5F7C9 was the canonical writer; bytes at that address are
+    /// stored as a literal string buffer (verified via memory viewer:
+    /// "main\0..." in main menu, "las_songs\0..." in LaS song-select, etc.).
+    ///
+    /// Returned as a (entryAddress, []) tuple so the existing FollowPointers
+    /// codepath in RSMemoryReader handles it uniformly — empty offsets means
+    /// the foreach loop is a no-op and the read happens at base+entry directly.
+    ///
+    /// KNOWN ENGINE BEHAVIOR (not a reader bug, do not "correct" here):
+    /// Rocksmith does not update this cell on pause→resume or pause→restart
+    /// transitions for ANY mode. The cell continues reading "*_pause" until
+    /// the user navigates to a menu or starts a different song. Consumers
+    /// needing actual play/pause state should consult `game_state` (the
+    /// SnifferState machine), which handles this correctly via the
+    /// timer-stall heuristic in Sniffer.UpdateState().
+    ///
+    /// EDITION SHIFTS (Beta / LaP): back-derived using the +0x3080 (Beta→Remastered)
+    /// and +0x4080 (Beta→LaP) shifts that every other pointer in this file uses.
+    /// That convention has held for all eight previously-mapped pointers/addresses,
+    /// so it is very likely correct here too — but the Beta and Learn_And_Play
+    /// values have not been independently verified. If those editions ever read
+    /// garbage / empty for gameStage with otherwise-working RockSniffer behavior,
+    /// check this address as the first suspect.
+    /// </remarks>
```
Clean up the AI ramblings here...

PoizenJam replied:
Cleaned up in https://github.com/PoizenJam/RockSnifferLib/commit/b338cfd5c24082d137dfc3b509c4ffb89418b2ed

kokolihapihvi commented on RSHelpers/MemoryOffsets.cs:
```diff
@@ -79,18 +79,65 @@ public static (int entryAddress, int[] offsets) GetArrangementHashPointer(RSEdit
-            RSEdition.Remastered_Learn_And_Play => (0x00F5C5AC + 0x4080, [0x18, 0x18, 0xC, 0x14]),
+            RSEdition.Remastered_Just_In_Case_We_Need_It_Beta => (0xF5F7C9 - 0x3080, []),
+            RSEdition.Remastered => (0xF5F7C9, []),
+            RSEdition.Remastered_Learn_And_Play => (0xF5F7C9 + 0x1000, []),
```
Format these addresses the same way as in the other functions for consistency.

PoizenJam replied:
Done in c2657d5.

kokolihapihvi commented on RSHelpers/MemoryOffsets.cs:
```diff
@@ -128,4 +175,234 @@ public static (int entryAddress, int[] offsets) GetScoreAttackNoteDataPointer(RS
+        // all seven previously-mapped pointers, so it is very likely correct here too —
+        // but the Beta and Learn_And_Play values have not been independently verified.
+        // If those editions ever read 0x00 for Path (with otherwise-working RockSniffer
+        // behavior), check this address as the first suspect.
```
Clean up AI ramblings...

PoizenJam replied:
Cleaned up in https://github.com/PoizenJam/RockSnifferLib/commit/b338cfd5c24082d137dfc3b509c4ffb89418b2ed

kokolihapihvi commented on RSHelpers/MemoryOffsets.cs:
```diff
@@ -128,4 +175,234 @@ public static (int entryAddress, int[] offsets) GetScoreAttackNoteDataPointer(RS
+    ///   0x01 → Lead
+    ///   0x02 → Rhythm
+    ///   0x04 → Bass
+    ///   anything else → Unknown
```
This description should probably be where the value exists after reading, or the value should be mapped to enum.

PoizenJam replied:
Handled by the enum work in ad89e07.

kokolihapihvi commented on RSHelpers/MemoryOffsets.cs:
```diff
@@ -128,4 +175,234 @@ public static (int entryAddress, int[] offsets) GetScoreAttackNoteDataPointer(RS
+    ///     buffer at +F5F7C9.
+    ///
+    /// IMPORTANT design note for state-machine consumers: because value=1
+    /// (tuner-from-pause) is distinct from value=2 (pause menu) but BOTH
+    /// represent "user is in a pause sub-flow," the correct test for
+    /// "currently paused" is mode != 0, not mode == 2. This eliminates the
+    /// tuner-from-pause edge case that complicated earlier pause-detection
+    /// designs — no asymmetric flag-entry / timer-exit gymnastics needed,
+    /// since the variable itself never lies about "we are in a pause overlay"
+    /// during tuner-from-pause.
+    ///
+    /// Returned as (entryAddress, []) tuple so the existing FollowPointers
+    /// codepath in RSMemoryReader handles it uniformly — empty offsets means
+    /// the foreach loop is a no-op and the read happens at base+entry
+    /// directly, matching the same pattern used for the gameStage static read.
+    ///
+    /// EDITION SHIFTS (Beta / LaP): back-derived using the +0x3080 (Beta→
+    /// Remastered) and +0x4080 (Beta→LaP) shifts that every other pointer
+    /// in this file uses. Convention has held for all eleven previously-mapped
+    /// pointers/addresses, so very likely correct — but Beta and Learn_And_Play
+    /// values have NOT been independently verified. If those editions read
+    /// constant zero across all pause states (otherwise-working RockSniffer
+    /// behavior), check this address as the first suspect.
+    ///
+    /// Discovery credit: kokolihapihvi (upstream RockSniffer maintainer)
+    /// pointed us at the surrounding memory region by sharing two named
+    /// candidate addresses from his RE project (MustBlockInputsDueToPauseMenu,
+    /// EnablePauseMenu). Both turned out to be dead in the current Remastered
+    /// build (consistently 0 across all states), but the neighborhood inspection
+    /// they prompted led directly to this find.
+    /// </remarks>
+    /// <param name="edition"></param>
+    /// <returns>A tuple of (entry address, pointer offsets) — offsets is empty
+    /// for a direct static read.</returns>
```
Clean up the AI ramblings...

PoizenJam replied:
Cleaned up in https://github.com/PoizenJam/RockSnifferLib/commit/b338cfd5c24082d137dfc3b509c4ffb89418b2ed

kokolihapihvi commented on RSHelpers/MemoryOffsets.cs:
```diff
@@ -128,4 +175,234 @@ public static (int entryAddress, int[] offsets) GetScoreAttackNoteDataPointer(RS
+        {
+            RSEdition.Remastered_Just_In_Case_We_Need_It_Beta => (0xF5F5FC - 0x3080, []),
+            RSEdition.Remastered => (0xF5F5FC, []),
+            RSEdition.Remastered_Learn_And_Play => (0xF5F5FC + 0x1000, []),
```
Format the addresses consistently with the other functions.

PoizenJam replied:
Done in c2657d5

kokolihapihvi commented on RSHelpers/MemoryOffsets.cs:
```diff
@@ -128,4 +175,234 @@ public static (int entryAddress, int[] offsets) GetScoreAttackNoteDataPointer(RS
+    /// binary read is cheaper than a 32-char ASCII string read with null-termination
+    /// scan; and (c) it removes the LaS code path's dependency on arrangement_hash,
+    /// which has historically been the more fragile of the two chains.
+    ///
+    /// CHAIN VALIDATION (PoizenJam, v0.6.8):
+    ///   - Survived full Rocksmith process restart with correct resolution to new
+    ///     heap address (singleton manager re-bind, same pattern observed for the
+    ///     existing arrangement_hash, song-timer, and note-data chains).
+    ///   - Survived multiple Nonstop entry/exit cycles, including round-trips
+    ///     through main menu and back into Nonstop.
+    ///   - Tracked correctly across multiple distinct songs and across distinct
+    ///     arrangements within the same song (Lead vs Rhythm vs Bass, plus
+    ///     bonus / alternate where present).
+    ///   - Output matches arrangement_hash in LaS gameplay (cross-validation
+    ///     against the previously-trusted chain).
+    ///
+    /// FORMAT NOTE: the GUID is returned as a string of 32 uppercase hex chars with
+    /// no separators, matching the example IDs observed in PSARC-side data (e.g.
+    /// EA94EC3F9817673B925B4997C4C0175C). Sniffer.cs cross-references readout.
+    /// arrangementID against songDetails.arrangements[].arrangementID using
+    /// case-sensitive string equality, so case normalization to upper is required
+    /// for downstream matching to succeed.
+    ///
+    /// EDITION SHIFTS (Beta / LaP): back-derived using the +0x3080 (Beta→Remastered)
+    /// and +0x4080 (Beta→LaP) shifts that every other pointer in this file uses.
+    /// Convention has held for all twelve previously-mapped pointers/addresses, so
+    /// very likely correct here too — but the Beta and Learn_And_Play values have
+    /// NOT been independently verified. If those editions ever read all-zero GUIDs
+    /// across all gameplay states (with otherwise-working RockSniffer behavior),
+    /// check this address as the first suspect.
+    /// </remarks>
+    /// <param name="edition"></param>
```
Clean up AI rambling.

PoizenJam replied:
Cleaned up in https://github.com/PoizenJam/RockSnifferLib/commit/b338cfd5c24082d137dfc3b509c4ffb89418b2ed

kokolihapihvi commented on RSHelpers/MemoryOffsets.cs:
```diff
@@ -128,4 +175,234 @@ public static (int entryAddress, int[] offsets) GetScoreAttackNoteDataPointer(RS
+        // (CE display order — outermost first, i.e. added last to the deepest pointer).
+        // Walk order (which FollowPointers expects) is the REVERSE: [0xB0, 0x18, 0x4, 0x84, 0x20].
+        // Same convention documented for GetCurrentPathPointer (v0.6.5 hotfix5) and
+        // observed for every other multi-offset chain in this file.
```
AI rambling again.

PoizenJam replied:
Cleaned up in https://github.com/PoizenJam/RockSnifferLib/commit/b338cfd5c24082d137dfc3b509c4ffb89418b2ed

kokolihapihvi commented on RSHelpers/RSMemoryReader.cs:
Clean up the AI rambling and version information from the comments in this file.

PoizenJam replied:
Cleaned up in https://github.com/PoizenJam/RockSnifferLib/commit/b338cfd5c24082d137dfc3b509c4ffb89418b2ed

kokolihapihvi commented on RSHelpers/RSMemoryReader.cs:
```diff
@@ -96,6 +115,196 @@ public RSMemoryReadout DoReadout()
+                {
+                    byte pathByte = MemoryHelper.ReadByteFromMemory(rsProcessHandle, pathAddr);
+                    readout.currentPathByte = pathByte;
+                    readout.currentPath = pathByte switch
```
This should probably be an enum, since there's only a few valid values.

PoizenJam replied:
Done in ad89e07:  RSPath { Unknown, Lead, Rhythm, Bass }

kokolihapihvi commented on RSHelpers/RSMemoryReader.cs:
```diff
@@ -172,6 +596,57 @@ private void ReadSongTimer(IntPtr timerAddress)
+        /// prior value (existing v0.6.5 "fail then retry on next poll" semantics
+        /// from the conditional IsValidArrangementHash assignment).
+        /// </summary>
+        private string ReadPlayArrIDFromMemory(IntPtr address)
```
Should probably be called `ReadGuidFromMemory`, since that's what it's doing.

PoizenJam replied:
Renamed to ReadGuidFromMemory in c2657d5.

kokolihapihvi commented on RSHelpers/RSMemoryReader.cs:
```diff
@@ -96,6 +115,196 @@ public RSMemoryReadout DoReadout()
+                if (pathAddr != IntPtr.Zero)
+                {
+                    byte pathByte = MemoryHelper.ReadByteFromMemory(rsProcessHandle, pathAddr);
+                    readout.currentPathByte = pathByte;
```
Probably don't need the raw byte either after mapping the value to an enum?

PoizenJam replied:
Dropped currentPathByte in ad89e07. The RSPath values equal the raw bytes (0x01/0x02/0x04), so the byte is recoverable by a cast if anything needs it.

kokolihapihvi commented on RSHelpers/RSMemoryReader.cs:
```diff
@@ -96,6 +115,196 @@ public RSMemoryReadout DoReadout()
+                if (pauseModeAddr != IntPtr.Zero)
+                {
+                    byte modeByte = MemoryHelper.ReadByteFromMemory(rsProcessHandle, pauseModeAddr);
+                    readout.pauseMenuMode = modeByte;
```
If the pause menu mode is important, map it to an enum as well with documented valid values.

PoizenJam replied:
Done in ad89e07: PauseMenuMode { None, SubOverlay, TopOverlay }

kokolihapihvi commented on RSHelpers/RSMemoryReadout.cs:
```diff
@@ -13,6 +15,67 @@ public class RSMemoryReadout
+        /// or "" (empty string) when the byte doesn't match a known value.
+        /// </summary>
+        public byte currentPathByte = 0;
+        public string currentPath = "";
```
Use enum instead since it's a set of a few valid values.

PoizenJam replied:
Done in ad89e07.

kokolihapihvi commented on RSHelpers/RSMemoryReadout.cs:
```diff
@@ -13,6 +15,67 @@ public class RSMemoryReadout
+        /// `pauseMenuMode != 0` test without that gating, suitable for addons
+        /// that don't care about main-menu Tools usage.
+        /// </summary>
+        public byte pauseMenuMode = 0;
```
Use enum.

PoizenJam replied:
Done in ad89e07

kokolihapihvi commented on RSHelpers/RSMemoryReadout.cs:
```diff
@@ -13,6 +15,67 @@ public class RSMemoryReadout
+        /// literal will break. None of the bundled addons in this repository
+        /// do that.
+        /// </summary>
+        [JsonConverter(typeof(StringEnumConverter))]
```
This is generally a good change in terms of usability, but all existing addons that use the mode will need to be updated. Unsure if this should be within the scope of this pull request.

PoizenJam replied:
Keeping it in scope. The RockSniffer PR already relies on the string form (sniffer-poller compares the member names) and updates all bundled addons, and the enum serialization (ad89e07) follows the same pattern. Third-party addons comparing integers need a one-line change; the release notes will call out all the JSON changes as breaking.

kokolihapihvi commented on RSHelpers/RSMode.cs:
```diff
@@ -6,11 +6,36 @@
+    ///
+    /// Integer values 0..3 are preserved from the pre-v0.6.8 enum for any
+    /// external consumer doing integer-based serialization. New values
+    /// (NONSTOPPLAY onwards) are appended.
```
AI rambling, unnecessary version history.

PoizenJam replied:
Cleaned up in https://github.com/PoizenJam/RockSnifferLib/commit/b338cfd5c24082d137dfc3b509c4ffb89418b2ed

kokolihapihvi commented on RSHelpers/RSMode.cs:
```diff
@@ -6,11 +6,36 @@
+        SESSION,        // 6 - sm_* (Session Mode)
+        LESSONS,        // 7 - ge_*, getuner, pregametuner
+        MENU            // 8 - titlescreen, profileselect, main, mainmenu, statsmenu,
+                        //     shop, contentpanelchord, sidelist, tonedesigner*
```
These comments feel unnecessary and are bound to go out of date and forgotten.
Names are probably descriptive enough, as the determining logic is documented in the read function.

PoizenJam replied:
Removed in c2657d5. The class doc points at DeriveModeFromGameStage, where the mapping is written out.

kokolihapihvi commented on Sniffing/ArrangementTuning.cs:
```diff
@@ -136,16 +136,43 @@ public class ArrangementTuning
+            var sb = new System.Text.StringBuilder();
+            for (int i = 0; i < stringCount; i++)
+            {
+                int semitone = ((StandardBase[i] + offsets[i]) % 12 + 12) % 12;
```
The % 12 + 12 % 12 looks redundant to me? Surely only the last % 12 is required?

PoizenJam replied:
This one is tricky and not immediately obvious. The two formulations _are_ equivalent for values:

(StandardBase[i] + offsets[i]) >= -12

However, the equivalence breaks down for values:

(StandardBase[i] + offsets[i]) < -12

These values exist in the dataset I was working in, so the initial clamp was necessary.

Friedfisch replied:
basic math says (not going into N-space math etc...): no. you cant get a negative number out of mod operation. maybe some article how that is possible link? wiki?

PoizenJam replied:
C# modulo/remainder operator is signed.

see worked example in the linked fiddle:
https://dotnetfiddle.net/p94SJ7

but do correct me if I’m mistaken somehow- the C programming language family is not my native working language.

Friedfisch replied:
makes no sense to get signed out of modulo. i can just read stuff. also not c# expert.

PoizenJam replied:
I don’t know what to say- other than the output of the fiddle I linked shows the two formulations are not mathematically equivalent in C#.

ffio1 replied:
What you said about it could be negative is possible, but when would a String tuning + offset ever be negative? What that essentially boils down to is a negative midi value which likely isn't supported in the game.

ffio1 replied:
Oh you're not using midi for standard base :/ That would solve your issue and make it easier to read imo.

PoizenJam replied:
This formula pertains to semitone offsets, not Hz or midi values.

StandardBase is a pitch class in semitones; The sum is an intermediate pitch-class index into NoteNames, not a note the game is expected to support, and not a midi value.

offsets aren’t bounded by game supported tuning. This renders whatever the CDLC declares- and semitone offsets can go lower than -12 from what I’ve observed. Several tunings in the dictionary in this same file go down to -14.

without the initial clamp,  C# signed remainder makes those cases a negative index into NoteNames which throws. I think one mod is cheap insurance against this.

kokolihapihvi commented on .gitattributes:
Normalizing line-endings is probably not a bad thing, but this seems excessive?

PoizenJam replied:
Removed from this PR in c2657d5

kokolihapihvi commented on RockSnifferLib.sln:
Shouldn't need a solution file for this project, remove it.

PoizenJam replied:
Removed in c2657d5.

kokolihapihvi commented on Sniffing/Sniffer.cs:
Clean up the AI rambling and version history comments in this file.

PoizenJam replied:
Cleaned up in https://github.com/PoizenJam/RockSnifferLib/commit/b338cfd5c24082d137dfc3b509c4ffb89418b2ed

kokolihapihvi commented on Sniffing/Sniffer.cs:
```diff
@@ -460,74 +796,664 @@ public void Stop()
+                    arrCount);
+            }
+
+            Logger.Log(
```
Logging events should probably be its own path instead of going through the general logger.

PoizenJam replied:
Agreed, these move to events.log as part of the event logging rework, if we're agreed on approach.

kokolihapihvi commented on Sniffing/Sniffer.cs:
```diff
@@ -460,74 +796,664 @@ public void Stop()
             fileSystemWatchers.Clear();
         }
 
+        /// <summary>
+        /// Update the state of the sniffer
+        /// </summary>
+        /// 
+        private void LogSongStartIfPossible()
```
Wrong doc block, questionable function name, it seems to do more than just log.
This function is massive and should probably be broken up to smaller parts for readability?

PoizenJam replied:
Agreed. I'll rename so the state change is explicit (BeginSongRun / EndSongRun, lined up with wherever the event naming lands in the L2/L4 thread), fix the copy-pasted doc block, and split each into guard, capture, fire, and log once settled on approach. 

Doing it with the events convergence so the rename and split happen once.

kokolihapihvi commented on Sniffing/Sniffer.cs:
```diff
@@ -460,74 +796,664 @@ public void Stop()
+        // 0.1s is well below any legitimate play time and well above the
+        // float-near-zero initialization flashes (e.g. 9.2515e-37) that can
+        // briefly appear during chart load.
+        private const float MIN_PROGRESS_SECONDS = 0.1f;
```
Move const to top of the file with the rest of the fields.

PoizenJam replied:
Done in c2657d5.

kokolihapihvi commented on Sniffing/Sniffer.cs:
```diff
@@ -460,74 +796,664 @@ public void Stop()
+            var noteData = snapshotReadout?.noteData ?? currentMemoryReadout.noteData;
+
+            // Build base log message
+            StringBuilder logMessage = new StringBuilder();
```
I don't like the mixed use of StringBuilder, interpolated strings, and classic concatenation in this file.
StringBuilder is efficient for when you're doing a lot of string building, but for something that happens once per song it probably doesn't matter.
Pick one style and use it consistently, interpolated strings might be best for this use case, you can break them across multiple lines for readability.

kokolihapihvi replied:
Interpolated strings inside `StringBuilder.Append` is acceptable since .NET 6 according to [this article](https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/6.0/stringbuilder-append-evaluation-order)

PoizenJam replied:
Standardized on interpolated strings in c2657d5

kokolihapihvi commented on Sniffing/Sniffer.cs:
```diff
@@ -460,74 +796,664 @@ public void Stop()
+            return true;
+        }
+
+        private void LogSongEnd(bool completed)
```
Revise function name, break into smaller chunks if needed, this function also modifies state, so just calling it Log... is misleading.

PoizenJam replied:
Same as previous; pending settled outstanding issues.

kokolihapihvi commented on Sniffing/Sniffer.cs:
```diff
@@ -460,74 +796,664 @@ public void Stop()
+                logMessage.Append($"HighestMultiplier={saData.HighestMultiplier};");
+            }
+
+            Logger.Log(logMessage.ToString());
```
This too should probably go through a different path than regular log messages.

PoizenJam replied:
Agreed, pending decision/discussion on events.log.

PoizenJam committed:
Comment cleanup: remove version-history narration, keep technical reference docs

PoizenJam committed:
Address review: remove .sln and .gitattributes, consistent address formatting, rename ReadGuidFromMemory, move const to fields, drop RSMode member comments, standardize interpolated strings

PoizenJam committed:
Address review: RSPath and PauseMenuMode enums with StringEnumConverter, drop currentPathByte, value docs moved onto enum members

PoizenJam committed:
Improve logging for restarted songs

- END payloads read from a rolling in-run readout snapshot instead of current
  memory; Restart zeroes noteData for the new attempt before the old attempt's
  end fires, which reported TotalNotes=0 / Accuracy=100
- Snapshot capture is timer-monotonic so the new attempt cannot overwrite the
  ended attempt's stats during sticky pause state
- Restart-vs-resume disambiguation in the pause-exit path via note-counter
  reset; quick restarts previously misdetected as resumes when the timer-0
  load transient fell between polls

PoizenJam committed:
Address review: document modulo clamp rationale (CDLC offsets below -12)

Also log 'Song Restarted!' on the timer<=initTime pause-exit path (previously
only the timer-moved path logged it, so restarts caught at the first-note seek
point were silent).

kokolihapihvi commented on RSHelpers/RSMemoryReader.cs:
```diff
@@ -201,8 +200,14 @@ public RSMemoryReadout DoReadout()
+                        0 => PauseMenuMode.None,
+                        1 => PauseMenuMode.SubOverlay,
+                        2 => PauseMenuMode.TopOverlay,
+                        _ => (PauseMenuMode)modeByte
```
What happens when you cast a byte to an enum, that's past the range of valid values?
Should there be an "unknown" case instead?

kokolihapihvi commented on Sniffing/Sniffer.cs:
```diff
@@ -700,7 +742,8 @@ private void LogSongStartIfPossible()
-            string currentPath = currentMemoryReadout?.currentPath;
+            string currentPath = (currentMemoryReadout != null && currentMemoryReadout.currentPath != RSPath.Unknown)
+                ? currentMemoryReadout.currentPath.ToString() : null;
             if (arrangement == null && !string.IsNullOrEmpty(currentPath) &&
```
Prefer plain if-statements over ternary for multiple conditions
